pkg/config/remote/store/store.go

```diff
@@ -0,0 +1,275 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package store
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	fmt "fmt"
+	"math"
+	"os"
+
+	"go.etcd.io/bbolt"
+
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo"
+	"github.com/DataDog/datadog-agent/pkg/security/log"
+)
+
+// Store allows storing configuration in a boltdb database
+type Store struct {
+	path          string
+	db            *bbolt.DB
+	maxBucketSize int
+	configBucket  string
+}
+
+// Close the store
+func (s *Store) Close() error {
+	return s.db.Close()
+}
+
+// keyFromVersion returns the boltdb key from a version.
+// boltdb entries are byte sorted so we need to convert it to big endian.
+func keyFromVersion(version uint64) []byte {
+	key := make([]byte, 8)
+	binary.BigEndian.PutUint64(key, version)
+	return key
+}
+
+// StoreConfig inserts a configuration into the database
+func (s *Store) StoreConfig(product string, config *pbgo.ConfigResponse) error {
+	if config == nil {
+		return errors.New("cannot store empty configuration")
+	}
+
+	err := s.db.Update(func(tx *bbolt.Tx) error {
+		configBucket, err := tx.CreateBucketIfNotExists([]byte(s.configBucket))
+		if err != nil {
+			return err
+		}
+
+		productBucket, err := configBucket.CreateBucketIfNotExists([]byte(product))
+		if err != nil {
+			return err
+		}
+
+		key := keyFromVersion(config.DirectoryTargets.Version)
+		data, err := json.Marshal(config)
+		if err != nil {
+			return err
+		}
+
+		return productBucket.Put(key, data)
+	})
+
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Prune %s bucket", product)
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		configBucket, err := s.getConfigBucket(tx)
+		if err != nil {
+			return err
+		}
+
+		productBucket := configBucket.Bucket([]byte(product))
+		if productBucket == nil {
+			return fmt.Errorf("could not find bucket %s: %w", product, bbolt.ErrBucketNotFound)
+		}
+
+		s.pruneBucket(productBucket)
+		return nil
+	})
+}
+
+func (s *Store) getConfigBucket(tx *bbolt.Tx) (*bbolt.Bucket, error) {
+	configBucket := tx.Bucket([]byte(s.configBucket))
+	if configBucket == nil {
+		return nil, fmt.Errorf("could not find bucket %s: %w", s.configBucket, bbolt.ErrBucketNotFound)
+	}
+	return configBucket, nil
+}
+
+// pruneBucket purges the bucket entries to keep only
+// maxBucketSize entries at maximum
+func (s *Store) pruneBucket(bucket *bbolt.Bucket) {
+	stats := bucket.Stats()
+	size := stats.KeyN
+	cursor := bucket.Cursor()
+	for k, _ := cursor.First(); k != nil && size > s.maxBucketSize; k, _ = cursor.Next() {
+		if err := cursor.Delete(); err != nil {
+			log.Errorf("failed to prune key %d: %s", k, err)
```

Comment:
It seems that `Delete` returns error when bucket is read-only, and subsequent calls to `Delete` can be expected to fail. Perhaps it would be better to break the loop here, so we don't spam the logs?

Reply:
True. I used a multierror and only log once all the errors